File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/Java/src/org/rlcommunity/rlglue/codec/RLGlue.java
/*
Copyright 2007 Brian Tanner
brian@tannerpages.com
http://brian.tannerpages.com

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
package org.rlcommunity.rlglue.codec;

import org.rlcommunity.rlglue.codec.types.Action;
import org.rlcommunity.rlglue.codec.types.Observation;
import org.rlcommunity.rlglue.codec.types.Observation_action;
import org.rlcommunity.rlglue.codec.types.Reward_observation_action_terminal;
import org.rlcommunity.rlglue.codec.types.Reward_observation_terminal;

/**
 * This is the main RLGlue interface class for Experiment Programs.
 * <p>This class is meant to be accessed statically, like:
 * <ul>
 * <li>RLGlue.RL_init()
 * <li>RLGlue.RL_episode(5)
 * <li>etc.
 * </ul>
 * <p>For future flexibility, we have made this class use an instance of type
 * RLGlueInterface.  This will allow RL-Viz and other project swap out the network
 * codec for a local (direct-compile) implementation without requiring any 
 * changes to the experiment program.
 * <p>Since this class offers some convenient extensions, the traditional RL-Glue
 * methods have all been labeled as 'RL-Glue Core Method'.
 * @author btanner
 */
public class RLGlue {

    private static RLGlueInterface instance = null;
    static boolean inited = false;
    static boolean currentEpisodeOver = true;

    /**
     * Non-Core Method.
     * <p> If you have an alternate RL-Glue engine (like the localGlue engine
     * from RL-Viz), you can use this method to specify the engine instead
     * of the network engine.
     * @param alternateGlueEngine The alternate glue engine.
     * @since 2.0
     */
    static public void setGlue(RLGlueInterface alternateGlueEngine) {
        resetGlueProxy();
        instance = alternateGlueEngine;
    }

    /**
     * Non-Core Method.
     * <p> If you wanted to run multiple independent experiments within a single
     * Experiment program, you could call this method in between.  This will disconnect
     * you from the glue.
     * <p> This has not been thoroughly tested : it is useful for RL-Viz.
     * @since 2.0
     */
    static public void resetGlueProxy() {
        instance = null;
        inited = false;
        currentEpisodeOver = true;
    }

    /**
     * Non-Core Method.
     * Has RL_init been called yet? (or since the last cleanup?)
     * @return Whether RL_init has been called.
     * @since 2.0
     */
    static public boolean isInited() {
        return inited;
    }

    /**
     *  RLGlue is only meant to be accessed statically, so this private 
     * constructor ensures we can never create an instance of RLGlue.
     */
    private RLGlue() {
    }

    private static void checkInstance() {
        if (instance == null) {
            instance = new NetGlue();
        }
    }

    /**
     * RL-Glue Core Method.
     */
    public static String RL_agent_message(String message) {
        checkInstance();
        if (message == null) {
            message = "";
        }
        String response = instance.RL_agent_message(message);
        if (response == null) {
            response = "";
        }
        return response;
    }

    /**
     * RL-Glue Core Method.
     */
    public static String RL_env_message(String message) {
        checkInstance();
        if (message == null) {
            message = "";
        }
        String response = instance.RL_env_message(message);
        if (response == null) {
            response = "";
        }
        return response;
    }

    /**
     * RL-Glue Core Method.
     */
    public static void RL_cleanup() {
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_cleanup() was called without matching RL_init() call previously.");
        }
        instance.RL_cleanup();
        inited = false;
    }

    /**
     * RL-Glue Core Method.
     */
    public static int RL_episode(int numSteps) {
        checkInstance();
        int terminationCondition = instance.RL_episode(numSteps);
        return terminationCondition;
    }


    /**
     * RL-Glue Core Method.
     */
    public static String RL_init() {
        checkInstance();
        if (inited) {
            System.err.println("-- Warning From RLGlue :: RL_init() was called more than once without RL_cleanup.");
        }
        String taskSpec = instance.RL_init();
        if (taskSpec == null) {
            taskSpec = "";
        }
        inited = true;
        currentEpisodeOver = true;
        return taskSpec;
    }

    /**
     * RL-Glue Core Method.
     */
    public static int RL_num_episodes() {
        checkInstance();
        return instance.RL_num_episodes();
    }

    /**
     * RL-Glue Core Method.
     */
    public static int RL_num_steps() {
        checkInstance();
        return instance.RL_num_steps();
    }

    /**
     * RL-Glue Core Method.
     */
    public static double RL_return() {
        checkInstance();
        return instance.RL_return();
    }


    /**
     * RL-Glue Core Method.
     */
    public static Observation_action RL_start() {
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_start() was called without RL_init().");
        }
        currentEpisodeOver = false;
        Observation_action returnObs = instance.RL_start();
        if (returnObs == null) {
            System.err.println("-- Warning From RLGlue :: RL_start() response was NULL, that should be impossible.");
            returnObs = new Observation_action();
        }
        return returnObs;
    }
    
    public static Observation RL_env_start(){
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_env_start() was called without RL_init().");
        }
        currentEpisodeOver = false;
        Observation returnObs = instance.RL_env_start();
        if (returnObs == null) {
            System.err.println("-- Warning From RLGlue :: RL_env_start() response was NULL, that should be impossible.");
            returnObs = new Observation();
        }
        return returnObs;
    }
    
    public static Action RL_agent_start(Observation theObservation){
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_agent_start() was called without RL_init().");
        }
        Action returnAct=instance.RL_agent_start(theObservation);
        if (returnAct == null) {
            System.err.println("-- Warning From RLGlue :: RL_agent_start() response was NULL, that should be impossible.");
            returnAct = new Action();
        }
        return returnAct;
    }

    /**
     * RL-Glue Core Method.
     */
    public static Reward_observation_action_terminal RL_step() {
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_step() was called without RL_init().");
        }
        Reward_observation_action_terminal stepResponse = instance.RL_step();
        if (stepResponse == null) {
            System.err.println("-- Warning From RLGlue :: RL_step() response was NULL, that should be impossible.");
            stepResponse = new Reward_observation_action_terminal();
        }

        currentEpisodeOver = (stepResponse.terminal == 1);
        return stepResponse;
    }
    /**
     * RL-Glue Core Method.
     */
    public static Reward_observation_terminal RL_env_step(Action theAction) {
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_env_step() was called without RL_init().");
        }
        Reward_observation_terminal stepResponse = instance.RL_env_step(theAction);
        if (stepResponse == null) {
            System.err.println("-- Warning From RLGlue :: RL_env_step() response was NULL, that should be impossible.");
            stepResponse = new Reward_observation_terminal();
        }
        currentEpisodeOver = stepResponse.isTerminal();
        return stepResponse;
    }
    
    
    public static Action RL_agent_step(double theReward, Observation theObservation){
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_agent_step() was called without RL_init().");
        }
        Action theAction=instance.RL_agent_step(theReward, theObservation);
        if (theAction == null) {
            System.err.println("-- Warning From RLGlue :: RL_agent_step() response was NULL, that should be impossible.");
            theAction = new Action();
        }
        return theAction;
    }
    public static void RL_agent_end(double theReward){
        checkInstance();
        if (!inited) {
            System.err.println("-- Warning From RLGlue :: RL_agent_end() was called without RL_init().");
        }
        instance.RL_agent_end(theReward);
    }

    /**
     * Non-Core Method.  Returns whether the current episode is active or not.
     * <p>Will return true BEFORE RL_start is called, and after a terminal response
     * has been returned from the environment (until next RL_start call).
     */
    public static boolean isCurrentEpisodeOver() {
        return currentEpisodeOver;
    }
}

File: Domains/InfiniteMario/15-rl-competition-2009/system/rl-glue/src/RL_glue.c
/* 
* Copyright (C) 2007, Andrew Butcher

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

* 
*  $Revision: 957 $
*  $Date: 2009-02-03 11:07:49 -0500 (Tue, 03 Feb 2009) $
*  $Author: brian@tannerpages.com $
*  $HeadURL: http://rl-glue.googlecode.com/svn/trunk/src/RL_glue.c $
* 
*/
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <rlglue/RL_glue.h>
#include <rlglue/Agent_common.h>
#include <rlglue/Environment_common.h>

#include <rlglue/utils/C/RLStruct_util.h>

#include <assert.h>


/*Brian, Sept 8, 2008 :
---------------------
I'm a little worried that some of my changes in this file 
(using stack variables for structs and returning them)
 might go terribly wrong*/

static const action_t *last_action=0;
static double total_reward = 0;
static int num_steps       = 0;
static int num_episodes    = 0;


const char* RL_init() {
  const char* task_spec;
  task_spec = env_init();
  agent_init(task_spec);

  total_reward=0;
  num_steps=0;
  num_episodes=0;
/* **WORRYSOME** */
	return task_spec;
}

const observation_action_t *RL_start() {
	const observation_t *last_state=0;
	/*Make this static so that it is safe to return it*/
	static observation_action_t oa;

	total_reward=0;
	num_steps=1;

	last_state = env_start();
	last_action = agent_start(last_state);

	oa.observation = last_state;
	oa.action = last_action;

/* **WORRYSOME** */
	return &oa;
}

const action_t* RL_agent_start(const observation_t* observation){
	return agent_start(observation);
}
const action_t* RL_agent_step(double reward, const observation_t* observation){
	return agent_step(reward, observation);
}
void RL_agent_end(double reward){
	agent_end(reward);
}
const observation_t* RL_env_start(){
		const observation_t *thisObservation=0;

		total_reward=0;
		num_steps=1;

		thisObservation = env_start();
		return thisObservation;
}
const reward_observation_terminal_t* RL_env_step(const action_t* action){
		const reward_observation_terminal_t *ro=0;
	  	double this_reward=0;

		__RL_CHECK_STRUCT(action)
	  	ro = env_step(action);
		__RL_CHECK_STRUCT(ro->observation)
	  	this_reward = ro->reward;

	  	total_reward += this_reward;

		 if (ro->terminal == 1) {
		   num_episodes += 1;
		 }
		 else {
			num_steps+=1;
		 }
		 return ro;
	}

const reward_observation_action_terminal_t *RL_step() {
	static reward_observation_action_terminal_t roa={0};
	const reward_observation_terminal_t *ro;
  	double this_reward=0;
	const observation_t *last_state;

	__RL_CHECK_STRUCT(last_action)
  	ro = env_step(last_action);
	__RL_CHECK_STRUCT(ro->observation)
  	this_reward = ro->reward;
  	last_state = ro->observation;
  
  	roa.reward = ro->reward;
  	roa.observation = ro->observation;
  	roa.terminal = ro->terminal;
  
  	total_reward += this_reward;

	/* Sept 28/08, The reason that we only increment stepcount if we're not terminal is that if an episode ends on 
	its first env_step, num_step will be 1 from env_start, but we don't want to go to num_step=2.*/
	 if (ro->terminal == 1) {
	   num_episodes += 1;
	   agent_end(this_reward);
	 }
	 else {
		num_steps+=1;
	   last_action = agent_step(this_reward,last_state);
	   __RL_CHECK_STRUCT(last_action)
	   roa.action = last_action;
	 }

/* **WORRYSOME** */
	 return &roa;
}

void RL_cleanup() {
  env_cleanup();
  agent_cleanup();
}

const char* RL_agent_message(const char* message) {
	const char *theAgentResponse=0;
	const char *messageToSend=0;

	if(message!=0){
		messageToSend=message;
	}else{
		messageToSend="";
	}

	theAgentResponse=agent_message(messageToSend);
	if(theAgentResponse==0){
		return "";
	}

	return theAgentResponse;
}

const char* RL_env_message(const char* message) {
	const char *theEnvResponse=0;
	const char *messageToSend=0;

	if(message!=0){
		messageToSend=message;
	}else{
		messageToSend="";
	}

	theEnvResponse=env_message(messageToSend);
	if(theEnvResponse==0){
		return "";
	}

	return theEnvResponse;
}

int RL_episode(const unsigned int maxStepsThisEpisode) {
	const reward_observation_action_terminal_t *rl_step_result=0;
	int isTerminal=0;

	
  	RL_start();
	/*RL_start sets steps to 1*/
  	for (; !isTerminal && (maxStepsThisEpisode == 0 ? 1 : num_steps < maxStepsThisEpisode); ) {
    	rl_step_result=RL_step();
		isTerminal=rl_step_result->terminal;
  	}

	/*Return the value of terminal to tell the caller whether the episode ended naturally or was cut off*/
	return isTerminal;
}

double RL_return() {
  return total_reward;
}

int RL_num_steps() {
  /* number of steps of the current or just completed episodes of run */
  return num_steps;
}

int RL_num_episodes() {
  return num_episodes;
}

